[Lens][Example] Testing playground for Lens embeddables

dej611 opened the pull request:
## Summary

This PR adds a new Lens embeddable example which I think it might be useful to debug and test/debug Lens configurations quickly.

The app will automatically identify fields per type within the default data view and map them into dimensions and metrics for the chart.

The interface is very similar to the Vega app, but the editor contains the Lens serialized `attributes`.

<img width="1222" alt="Screenshot 2022-02-09 at 12 22 17" src="https://user-images.githubusercontent.com/924948/153192704-ad91affb-7e92-4b89-9208-8e3f5da941b8.png">

There are presets that can be used to quick start a particular configuration, and a custom configuration can be saved as well as preset (it is not stored, so it will be available as long as the page is open).

<img width="439" alt="Screenshot 2022-02-09 at 12 22 54" src="https://user-images.githubusercontent.com/924948/153192957-41d5a7cf-0290-43f8-ab95-ef34500aa75a.png">

<img width="346" alt="Screenshot 2022-02-09 at 12 23 00" src="https://user-images.githubusercontent.com/924948/153192969-6a03e8ba-fa60-438e-82d2-a754c33f0d59.png">

Editing the serialized attributes will automatically render the new chart:
![testing_playground_edit_so](https://user-images.githubusercontent.com/924948/153193093-ca729a3c-3b69-480e-b2ee-9579a8e6c41a.gif)

### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

elasticmachine commented:
Pinging @elastic/kibana-vis-editors @elastic/kibana-vis-editors-external (Team:VisEditors)

mbondyra commented on x-pack/examples/testing_embedded_lens/public/app.tsx:
```diff
@@ -0,0 +1,679 @@
+                        }}
+                        isDisabled={isColorDisabled}
+                      >
+                        Change color
```
I noticed that you use this button to indicate when there's a loading state. I think it's enough for this demo, but it makes the whole panel jump down because of the button width changes. What if you change this copy to:
```suggestion
                        {isLoading ? 'Loading' : 'Change color'}
```
then the button stays the same width and it stays in place. 

Uploading jump.mp4…

mbondyra commented on x-pack/examples/testing_embedded_lens/public/app.tsx:
```diff
@@ -0,0 +1,679 @@
+                        isLoading={isLoading}
+                        onClick={() => {
+                          // eslint-disable-next-line no-bitwise
+                          const newColor = '#' + ((Math.random() * 0xffffff) << 0).toString(16);
```
nit: you can also do the below to avoid the bitwise eslint warning :) 
```suggestion
                          const newColor = `#${Math.floor(Math.random()*16777215).toString(16)}`;
```

mbondyra commented:
I am not sure if I read this completely right, but the state for gauges is not properly presented/cleaned. It still has the bucket column, and gauge shouldn't have any. 
<img width="1035" alt="Screenshot 2022-02-10 at 21 22 46" src="https://user-images.githubusercontent.com/4283304/153490442-63ab58ea-db9b-4e3b-8942-0cecd7d94966.png">

mbondyra reviewed (commented):
There is one more issue I found now:
1. When modifying the code inline and creating invalid state, the page blows up (here typing " but anything that makes JSON incorrect will do)

https://user-images.githubusercontent.com/4283304/153595429-6a6cbe02-a4a9-422c-9ca6-d423f53c8ff1.mp4


I am also wondering why do we create a new plugin for testing and have one for embedding, and they are almost the same? Couldn't we merge them into one so we don't have to support 2 of them? @flash1293 what's your opinion here?

flash1293 commented:
I don't have much concern about having two plugins here, one kept simple show-casing how to integrate it into a new app and one as a "playground" kind of thing. If anything I think this allows us to simplify the other example a bit (like enable/disable actions, open in new tab and so on - doesn't seem necessary to get started)

dej611 commented:
I could remove the actions buttons there and move them here. It makes sense to have a simpler examples for development showcase tbh.

dej611 commented:
Latest updates:

* Removed some extra buttons from the developer Lens example and moved in here
* Fix the problem for invalid spec and show now an error callout:
<img width="495" alt="Screenshot 2022-02-15 at 13 14 01" src="https://user-images.githubusercontent.com/924948/154060303-9c35f5d4-fddc-46ab-82af-dc7bbdf57d76.png">

flash1293 commented:
Changing the color doesn't work anymore - it seems like the new scheme of protecting against JSON errors broke the update

mbondyra reviewed (approved):
Works great! Approving 🆗
